File: 프로그래머스/lv2/49994. 방문 길이/방문 길이.py
# 방문체크를 하면서 위치를 이동한다
def solution(dirs):
    answer = 0
    visited = [[[False] * 11 for _ in range(11)] for _ in range(4)]
    dict = {"U": 0, "D": 1, "R": 2, "L": 3}
    dy = [-1, 1, 0, 0]
    dx = [0, 0, 1, -1]

    y, x = 5, 5
    for d in dirs:
        i = dict[d]

        ny = y + dy[i]
        nx = x + dx[i]

        # 1 좌표상으로 방문하지 못할때
        if ny < 0 or ny > 10 or nx < 0 or nx > 10:
            continue

        # 2 좌표상으로 갈수는 있고 이전에 방문한 적이 없을때
        j = i
        if i % 2 == 0:
            j += 1
        else:
            j -= 1

        if not visited[i][ny][nx] and not visited[j][y][x]:
            visited[i][ny][nx] = True
            answer += 1

        y, x = ny, nx

    return answer

File: 프로그래머스/lv2/49994. 방문 길이/방문 길이.java
import java.util.*;

class Solution {
    public int solution(String dirs) {
        int answer = 0;
        
        boolean visited[][][] = new boolean[4][11][11];
        Map<Character, Integer> map = new HashMap<Character, Integer>();
        map.put('U', 0);
        map.put('D', 1);
        map.put('R', 2);
        map.put('L', 3);
        
        int[] dy = {-1, 1, 0, 0};
        int[] dx = {0, 0, 1, -1};
        
        int y = 5;
        int x = 5;
        
        for(int d = 0; d < dirs.length(); d++) {
            int i = map.get(dirs.charAt(d));
            
            int ny = y + dy[i];
            int nx = x + dx[i];
            
            if(ny < 0 || ny > 10 || nx < 0 || nx > 10) {
                continue;
            }
            
            int j = i;
            if (i % 2 == 0){
                j++;
            } else {
                j--;
            }
            
            if (!visited[i][ny][nx] && !visited[j][y][x]) {
                visited[i][ny][nx] = true;
                answer++;
            }
            y = ny;
            x = nx;
        }
        return answer;
    }
}